# 📊 DATA 201 – CAPSTONE PROJECT

## 1. Introduction

### 1.1 Background

Motor vehicle collisions are an important public safety issue in New York City. Analyzing collision data can help identify patterns related to location, time, contributing factors, vehicle types, injuries, and fatalities.

### 1.2 Motivation

The purpose of this project is to use NYC motor vehicle collision data to identify patterns that may help better understand when, where, and under what circumstances collisions occur.

### 1.3 Research Questions

1. **Which NYC borough has the highest number of motor vehicle collisions?**

2. **At what times and on which days do motor vehicle collisions occur most frequently?**

3. **Which contributing factors and vehicle types are most associated with injury or fatal crashes?**



## 2. Data Source

### 2.1 Dataset Description

This project uses the **Motor Vehicle Collisions – Crashes** dataset from NYC Open Data. The dataset contains information about motor vehicle collisions reported by the New York City Police Department (NYPD), including crash date and time, borough, injuries, fatalities, contributing factors, and vehicle types.


### 2.2 Import Libraries

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

### 2.3 Load Dataset

The dataset was loaded directly from NYC Open Data using its CSV API.

In [ ]:
df = pd.read_csv(
"https://data.cityofnewyork.us/api/v3/views/h9gi-nx95/query.csv"
)


/tmp/ipykernel_2778/339635639.py:1: DtypeWarning:

Columns (7) have mixed types. Specify dtype option on import or set low_memory=False.



## 3. Understanding the Data

This section provides a quick overview of the dataset before cleaning and analysis.

### 3.1 Dataset Preview

In [ ]:
df.head()

,:id,:version,:created_at,:updated_at,crash_date,crash_time,borough,zip_code,latitude,longitude,...,contributing_factor_vehicle_2,contributing_factor_vehicle_3,contributing_factor_vehicle_4,contributing_factor_vehicle_5,collision_id,vehicle_type_code1,vehicle_type_code2,vehicle_type_code_3,vehicle_type_code_4,vehicle_type_code_5
0,row-rsc3.wgwr-zpi9,rv-qukg_54rg~e45q,2021-09-14T22:52:22.882Z,2021-09-14T22:52:26.672Z,2021-09-11T00:00:00.000,2:39,NaN,NaN,NaN,NaN,...,Unspecified,NaN,NaN,NaN,4455765,Sedan,Sedan,NaN,NaN,NaN
1,row-t9dc~q8a7~jwk4,rv-vmia.j6tc.thwm,2022-03-29T22:52:48.720Z,2022-03-29T22:52:49.994Z,2022-03-26T00:00:00.000,11:45,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,4513547,Sedan,NaN,NaN,NaN,NaN
2,row-h8ny_qxbv_ezwp,rv-x4h8~yx6x.vdk4,2023-11-06T23:52:10.201Z,2023-11-06T23:52:18.104Z,2023-11-01T00:00:00.000,1:29,BROOKLYN,11230.0,40.62179,-73.970024,...,Unspecified,Unspecified,NaN,NaN,4675373,Moped,Sedan,Sedan,NaN,NaN
3,row-gxrs-hm78-w9gs,rv-fazg.rwsa.9gjd,2022-07-04T22:51:56.069Z,2022-07-04T22:52:04.895Z,2022-06-29T00:00:00.000,6:55,NaN,NaN,NaN,NaN,...,Unspecified,NaN,NaN,NaN,4541903,Sedan,Pick-up Truck,NaN,NaN,NaN
4,row-58i4.zk9q.sxqs,rv-4bks~69au~m7kz,2022-09-26T22:52:05.596Z,2022-09-26T22:52:11.092Z,2022-09-21T00:00:00.000,13:21,NaN,NaN,NaN,NaN,...,Unspecified,NaN,NaN,NaN,4566131,Station Wagon/Sport Utility Vehicle,NaN,NaN,NaN,NaN


### 3.2 Dataset Structure

In [ ]:
print("Rows:", df.shape[0])
print("Columns:", df.shape[1])

df.info()

Rows: 2269187
Columns: 33
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2269187 entries, 0 to 2269186
Data columns (total 33 columns):
 #   Column                         Dtype  
---  ------                         -----  
 0   :id                            object 
 1   :version                       object 
 2   :created_at                    object 
 3   :updated_at                    object 
 4   crash_date                     object 
 5   crash_time                     object 
 6   borough                        object 
 7   zip_code                       object 
 8   latitude                       float64
 9   longitude                      float64
 10  location                       object 
 11  on_street_name                 object 
 12  off_street_name                object 
 13  cross_street_name              object 
 14  number_of_persons_injured      float64
 15  number_of_persons_killed       float64
 16  number_of_pedestrians_injured  int64  
 17  number_of_pedestrian

### 3.3 Summary Statistics

In [ ]:
df.describe()

,latitude,longitude,number_of_persons_injured,number_of_persons_killed,number_of_pedestrians_injured,number_of_pedestrians_killed,number_of_cyclist_injured,number_of_cyclist_killed,number_of_motorist_injured,number_of_motorist_killed,collision_id
count,2.028381e+06,2.028381e+06,2.269169e+06,2.260274e+06,2.269187e+06,2.269187e+06,2.269187e+06,2.269187e+06,2.269187e+06,2.269187e+06,2.269187e+06
mean,4.057145e+01,-7.364972e+01,3.333172e-01,1.600248e-03,6.073497e-02,7.954391e-04,2.967406e-02,1.295618e-04,2.380707e-01,6.460464e-04,3.300453e+06
std,2.487895e+00,4.609250e+00,7.187101e-01,4.221063e-02,2.521715e-01,2.882608e-02,1.720208e-01,1.142044e-02,6.799336e-01,2.776300e-02,1.511531e+06
min,0.000000e+00,-2.013600e+02,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,2.200000e+01
25%,4.066724e+01,-7.397446e+01,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,3.203418e+06
50%,4.072016e+01,-7.392653e+01,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,3.770912e+06
75%,4.076963e+01,-7.386636e+01,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,4.338406e+06
max,4.334444e+01,0.000000e+00,4.300000e+01,8.000000e+00,2.700000e+01,6.000000e+00,4.000000e+00,2.000000e+00,4.300000e+01,5.000000e+00,4.905771e+06


In [ ]:
df[
    [
        "number_of_persons_injured",
        "number_of_persons_killed"
    ]
].describe()

,number_of_persons_injured,number_of_persons_killed
count,2.269169e+06,2.260274e+06
mean,3.333172e-01,1.600248e-03
std,7.187101e-01,4.221063e-02
min,0.000000e+00,0.000000e+00
25%,0.000000e+00,0.000000e+00
50%,0.000000e+00,0.000000e+00
75%,0.000000e+00,0.000000e+00
max,4.300000e+01,8.000000e+00


## 4. Data Cleaning and Transformation




### 4.1 Select Relevant Columns

* Before cleaning the dataset, we first examine missing values, duplicate records, data types, and unique values. This helps us understand what problems exist and decide how to handle them correctly.

In [ ]:
columns_to_use = [
    "crash_date",
    "crash_time",
    "borough",
    "number_of_persons_injured",
    "number_of_persons_killed",
    "number_of_pedestrians_injured",
    "number_of_pedestrians_killed",
    "number_of_cyclist_injured",
    "number_of_cyclist_killed",
    "number_of_motorist_injured",
    "number_of_motorist_killed",
    "contributing_factor_vehicle_1",
    "contributing_factor_vehicle_2",
    "vehicle_type_code1",
    "vehicle_type_code2",
    "collision_id"
]

df = df[columns_to_use].copy()

### 4.2 Missing Values

In [ ]:
df.isnull().sum()

,0
crash_date,0
crash_time,0
borough,691375
number_of_persons_injured,18
number_of_persons_killed,8913
number_of_pedestrians_injured,0
number_of_pedestrians_killed,0
number_of_cyclist_injured,0
number_of_cyclist_killed,0
number_of_motorist_injured,0


In [ ]:
missing_values = df.isnull().sum()

missing_percentage = (
    df.isnull().sum() / len(df) * 100
).round(2)

missing_data = pd.DataFrame({
    "Missing Values": missing_values,
    "Percentage": missing_percentage
})

missing_data

,Missing Values,Percentage
crash_date,0,0.00
crash_time,0,0.00
borough,691375,30.47
number_of_persons_injured,18,0.00
number_of_persons_killed,8913,0.39
number_of_pedestrians_injured,0,0.00
number_of_pedestrians_killed,0,0.00
number_of_cyclist_injured,0,0.00
number_of_cyclist_killed,0,0.00
number_of_motorist_injured,0,0.00


### 4.3 Duplicates

In [ ]:
print("Duplicate rows:", df.duplicated().sum())
print("Duplicate collision IDs:", df["collision_id"].duplicated().sum())

Duplicate rows: 0
Duplicate collision IDs: 0


In [ ]:
df_clean = df.copy()

### 4.4 Data Types

In [ ]:
df_clean["crash_date"] = pd.to_datetime(
    df_clean["crash_date"],
    errors="coerce"
)

In [ ]:
df_clean["crash_hour"] = pd.to_numeric(
    df_clean["crash_time"].str.split(":").str[0],
    errors="coerce"
).astype("Int64")

In [ ]:
df_clean[
    ["crash_time", "crash_hour"]
].head()

,crash_time,crash_hour
0,2:39,2
1,11:45,11
2,1:29,1
3,6:55,6
4,13:21,13


In [ ]:
df_clean["crash_hour"].value_counts().sort_index()

,count
crash_hour,
0,75806
1,40498
2,31269
3,27571
4,31321
5,33878
6,51553
7,70581
8,124228


### 4.5 Categorical Data Cleaning

In [ ]:
# Borough
df_clean["borough"] = (
    df_clean["borough"]
    .astype("string")
    .str.strip()
    .str.upper()
)

# Contributing factors
factor_columns = [
    "contributing_factor_vehicle_1",
    "contributing_factor_vehicle_2"
]

for column in factor_columns:
    df_clean[column] = (
        df_clean[column]
        .astype("string")
        .str.strip()
    )

# Vehicle types
vehicle_columns = [
    "vehicle_type_code1",
    "vehicle_type_code2"
]

for column in vehicle_columns:
    df_clean[column] = (
        df_clean[column]
        .astype("string")
        .str.strip()
    )

### 4.6 Injury and Fatality Validation

In [ ]:
# Total Persons Injured
# Total Persons Killed
# Pedestrians, Cyclists, and Motorists

calculated_injured = (
    df_clean["number_of_pedestrians_injured"]
    + df_clean["number_of_cyclist_injured"]
    + df_clean["number_of_motorist_injured"]
)

calculated_killed = (
    df_clean["number_of_pedestrians_killed"]
    + df_clean["number_of_cyclist_killed"]
    + df_clean["number_of_motorist_killed"]
)

df_clean["number_of_persons_injured"] = (
    df_clean["number_of_persons_injured"]
    .fillna(calculated_injured)
)

df_clean["number_of_persons_killed"] = (
    df_clean["number_of_persons_killed"]
    .fillna(calculated_killed)
)

In [ ]:
columns_to_drop = [
    "number_of_pedestrians_injured",
    "number_of_pedestrians_killed",
    "number_of_cyclist_injured",
    "number_of_cyclist_killed",
    "number_of_motorist_injured",
    "number_of_motorist_killed"
]

df_clean = df_clean.drop(columns=columns_to_drop)

In [ ]:
df_clean.columns

Index(['crash_date', 'crash_time', 'borough', 'number_of_persons_injured',
       'number_of_persons_killed', 'contributing_factor_vehicle_1',
       'contributing_factor_vehicle_2', 'vehicle_type_code1',
       'vehicle_type_code2', 'collision_id', 'crash_hour'],
      dtype='object')

### 4.7 Create Analysis Variables

In [ ]:
df_clean["year"] = df_clean["crash_date"].dt.year
df_clean["month"] = df_clean["crash_date"].dt.month_name()
df_clean["day_of_week"] = df_clean["crash_date"].dt.day_name()

df_clean["crash_severity"] = "No Injury"

df_clean.loc[
    df_clean["number_of_persons_injured"] > 0,
    "crash_severity"
] = "Injury"

df_clean.loc[
    df_clean["number_of_persons_killed"] > 0,
    "crash_severity"
] = "Fatal"

In [ ]:
display(
    df_clean[
        [
            "crash_date",
            "year",
            "month",
            "day_of_week",
            "crash_time",
            "crash_hour",
            "crash_severity"
        ]
    ].head()
)


,crash_date,year,month,day_of_week,crash_time,crash_hour,crash_severity
0,2021-09-11,2021,September,Saturday,2:39,2,Injury
1,2022-03-26,2022,March,Saturday,11:45,11,Injury
2,2023-11-01,2023,November,Wednesday,1:29,1,Injury
3,2022-06-29,2022,June,Wednesday,6:55,6,No Injury
4,2022-09-21,2022,September,Wednesday,13:21,13,No Injury


### 4.8 Final Dataset Check

In [ ]:
print("Dataset shape:", df_clean.shape)

print("\nMissing values:")
print(df_clean.isnull().sum())

print("\nDuplicate rows:")
print(df_clean.duplicated().sum())

#display(df_clean.head())

Dataset shape: (2269187, 15)

Missing values:
crash_date                            0
crash_time                            0
borough                          691375
number_of_persons_injured             0
number_of_persons_killed              0
contributing_factor_vehicle_1      8202
contributing_factor_vehicle_2    368580
vehicle_type_code1                17093
vehicle_type_code2               463159
collision_id                          0
crash_hour                            0
year                                  0
month                                 0
day_of_week                           0
crash_severity                        0
dtype: int64

Duplicate rows:
0


## 5 Data analisis 🔢

### 5.1 Research Question 1

**Which NYC borough has the highest number of motor vehicle collisions?**

The number of collisions was grouped by borough to identify which area of New York City reported the highest number of motor vehicle crashes.

In [ ]:
# Count collisions by borough
borough_collisions = (
    df_clean["borough"]
    .value_counts()
    .sort_values(ascending=False)
)

borough_collisions

,count
borough,
BROOKLYN,506806
QUEENS,422409
MANHATTAN,348485
BRONX,234088
STATEN ISLAND,66024


In [ ]:
# Count collisions by borough
borough_collisions = (
    df_clean["borough"]
    .dropna()
    .value_counts()
    .reset_index()
)

borough_collisions.columns = ["borough", "collisions"]

# Interactive bar chart
fig = px.bar(
    borough_collisions,
    x="borough",
    y="collisions",
    color="collisions",
    text="collisions",
    title="Motor Vehicle Collisions by Borough",
    labels={
        "borough": "Borough",
        "collisions": "Number of Collisions"
    }
)

fig.update_traces(textposition="outside")
fig.update_layout(
    xaxis_title="Borough",
    yaxis_title="Number of Collisions",
    showlegend=False
)

fig.show()

In [ ]:
top_borough = borough_collisions.iloc[0]["borough"]
top_collisions = borough_collisions.iloc[0]["collisions"]

print("Borough with the most collisions:", top_borough)
print("Number of collisions:", top_collisions)

Borough with the most collisions: BROOKLYN
Number of collisions: 506806


**Interpretation:**  

Among collisions with a reported borough, **Brooklyn** had the highest number of motor vehicle collisions, with **506,806 recorded crashes**.

### 5.2 Research Question 2

**At what times and on which days do motor vehicle collisions occur most frequently?**

The number of collisions was analyzed by hour of the day and day of the week to identify when crashes occur most frequently.

In [ ]:
# Collisions by hour
hour_collisions = (
    df_clean["crash_hour"]
    .value_counts()
    .sort_index()
)

hour_collisions



,count
crash_hour,
0,75806
1,40498
2,31269
3,27571
4,31321
5,33878
6,51553
7,70581
8,124228


In [ ]:
# Collisions by hour
hour_collisions = (
    df_clean["crash_hour"]
    .dropna()
    .value_counts()
    .sort_index()
    .reset_index()
)

hour_collisions.columns = ["crash_hour", "collisions"]

# Interactive line chart
fig = px.line(
    hour_collisions,
    x="crash_hour",
    y="collisions",
    markers=True,
    title="Motor Vehicle Collisions by Hour of Day",
    labels={
        "crash_hour": "Hour of Day",
        "collisions": "Number of Collisions"
    }
)

fig.update_layout(
    xaxis=dict(tickmode="linear", dtick=1),
    xaxis_title="Hour of Day",
    yaxis_title="Number of Collisions"
)

fig.show()

In [ ]:
day_order = [
    "Monday", "Tuesday", "Wednesday",
    "Thursday", "Friday", "Saturday", "Sunday"
]

day_collisions = (
    df_clean["day_of_week"]
    .value_counts()
    .reindex(day_order)
    .reset_index()
)

day_collisions.columns = ["day_of_week", "collisions"]


In [ ]:
# Interactive bar chart
fig = px.bar(
    day_collisions,
    x="day_of_week",
    y="collisions",
    color="collisions",
    text="collisions",
    title="Motor Vehicle Collisions by Day of Week",
    labels={
        "day_of_week": "Day of Week",
        "collisions": "Number of Collisions"
    }
)

fig.update_traces(textposition="outside")
fig.update_layout(
    xaxis_title="Day of Week",
    yaxis_title="Number of Collisions",
    showlegend=False
)

fig.show()

In [ ]:
print("Hour with most collisions:", hour_collisions.idxmax())
print("Day with most collisions:", day_collisions.idxmax())

Hour with most collisions: crash_hour    23
collisions    16
dtype: int64
Day with most collisions: day_of_week    2
collisions     4
dtype: int64


**Interpretation:**  
Motor vehicle collisions occurred most frequently around **4:00 PM (16:00)**, while **Friday** had the highest number of recorded crashes.

### 5.3 Research Question 3

**Which contributing factors and vehicle types are most common among injury or fatal crashes?**

This analysis focuses only on crashes classified as `"Injury"` or `"Fatal"` to identify the most frequently recorded contributing factors and vehicle types.

In [ ]:
# Keep only injury or fatal crashes
severe_crashes = df_clean[
    df_clean["crash_severity"].isin(["Injury", "Fatal"])
]

# Exclude "Unspecified"
top_factors = (
    severe_crashes[
        severe_crashes["contributing_factor_vehicle_1"] != "Unspecified"
    ]["contributing_factor_vehicle_1"]
    .value_counts()
    .head(10)
    .reset_index()
)

top_factors.columns = ["factor", "collisions"]

In [ ]:
# Interactive horizontal bar chart
fig = px.bar(
    top_factors.sort_values("collisions"),
    x="collisions",
    y="factor",
    orientation="h",
    color="collisions",
    text="collisions",
    title="Top Contributing Factors in Injury or Fatal Crashes",
    labels={
        "factor": "Contributing Factor",
        "collisions": "Number of Crashes"
    }
)

fig.update_traces(textposition="outside")
fig.update_layout(
    xaxis_title="Number of Crashes",
    yaxis_title="Contributing Factor",
    showlegend=False
)

fig.show()

In [ ]:
#Top vehicles
top_vehicles = (
    severe_crashes["vehicle_type_code1"]
    .value_counts()
    .head(10)
    .reset_index()
)

top_vehicles.columns = ["vehicle_type", "collisions"]

In [ ]:
# Interactive horizontal bar chart
fig = px.bar(
    top_vehicles.sort_values("collisions"),
    x="collisions",
    y="vehicle_type",
    orientation="h",
    color="collisions",
    text="collisions",
    title="Top Vehicle Types in Injury or Fatal Crashes",
    labels={
        "vehicle_type": "Vehicle Type",
        "collisions": "Number of Crashes"
    }
)

fig.update_traces(textposition="outside")
fig.update_layout(
    xaxis_title="Number of Crashes",
    yaxis_title="Vehicle Type",
    showlegend=False
)

fig.show()

In [ ]:
print(
    "Most common contributing factor:",
    top_factors.iloc[0]["factor"]
)

print(
    "Most common vehicle type:",
    top_vehicles.iloc[0]["vehicle_type"]
)

Most common contributing factor: Driver Inattention/Distraction
Most common vehicle type: Sedan


**Interpretation:**  
The analysis shows that **Driver Inattention/Distraction** is the most common contributing factor among injury or fatal crashes, while **Sedan** is the most frequently recorded vehicle type in these collisions.

## 6 Key Findings 🔐

The analysis identified three main findings:

1. **BROKLYN** had the highest number of reported motor vehicle collisions.
2. Collisions occurred most frequently around **16**, and **FRIDAY** had the highest number of crashes.
3. **Driver Inattention/Distraction** was the most frequently recorded contributing factor among injury or fatal crashes, although it does not identify a specific cause. **Sedan** was the most frequently recorded vehicle type in these crashes.

## 7 Conclusion

The analysis of NYC motor vehicle collisions revealed clear patterns related to location, time, contributing factors, and vehicle types.

### 7.1 Answers to the Research Questions

- **Borough:** Among crashes with a reported borough, **Brooklyn** recorded the highest number of motor vehicle collisions.
- **Time and Day:** Collisions occurred most frequently around **4:00 PM**, and **Friday** had the highest number of recorded crashes.
- **Severe Crashes:** Among injury or fatal crashes, **Driver Inattention/Distraction** was the most frequently recorded specific contributing factor after excluding `"Unspecified"`, while **Sedan** was the most frequently recorded vehicle type.

### 7.2 Recommendation

Traffic safety efforts could focus on high-collision periods and locations, while improving the completeness and consistency of contributing-factor reporting would support more reliable crash analysis.

### 8.3 Limitations

This analysis has several limitations:

- A substantial number of records are missing borough information.
- Many contributing-factor records are classified as `"Unspecified"`.
- Vehicle-type categories are not always standardized, which may split similar vehicles into different categories.

## 8 References

- NYC Open Data. **Motor Vehicle Collisions – Crashes**. New York City Police Department (NYPD).  
  Dataset: https://data.cityofnewyork.us/Public-Safety/Motor-Vehicle-Collisions-Crashes/h9gi-nx95

- NYC Open Data API.  
  https://data.cityofnewyork.us/api/v3/views/h9gi-nx95/query.csv

- Pandas Documentation.  
  https://pandas.pydata.org/docs/

- Matplotlib Documentation.  
  https://matplotlib.org/stable/